blog/02-18-2023-Build-an-NFT-Full-Stack-Monster-War-Game-on-The-Celo-Blockchain/Build-an-NFT-Full-Stack-Monster-War-Game-on-The-Celo-Blockchain.md

```diff
@@ -0,0 +1,998 @@
+            _tokenId,
+            payable(msg.sender),
+            _powerValue,
+            _name
+        );
+        allNFTs++;
+        minters[msg.sender] = true;
+    }
+```
+
+The `AddNFT()` function adds the NFT to the game. The function takes a `tokenId` and a name as parameters. It then creates a new NFT struct with the given parameters and adds it to the internal nfts mapping. It also adds the player to the minters mapping and increments the `allNFTs` variable. 
+
+```solidity
+function swallowNFT(uint _index) external hasmint(msg.sender) canSwallow(msg.sender, _index){
+	       require(msg.sender != nfts[_index].owner, "can't swallow your own nft");         
+           _transfer(nfts[_index].owner, msg.sender, nfts[_index].tokenId);
+           playerpowervalue[nfts[_index].owner] -= nfts[_index].powerValue;
+           playerpowervalue[msg.sender] += nfts[_index].powerValue;
+           nfts[_index].owner = payable(msg.sender);
+  }
+```
+
+The `SwallowNFT()` function allows a player to swallow another player’s NFT. This function takes an index as a parameter which is used to specify the NFT to be swallowed. The function is protected by two modifiers. `hasmint` and `canSwallow`. The `hasmint` modifier checks if the address has minted an NFT and the `canSwallow` modifier checks if the power value of the attacker is greater than the power value of the owner. If both the modifiers pass, the function transfers the NFT from the owner to the attacker, decreases the power value of the owner and increases the power value of the attacker. 
+
+```solidity
+ function upgradeNFT(uint _index) external payable{
+        require(msg.sender == nfts[_index].owner, "you cant upgrade this nft");
+         payable(owner()).transfer(msg.value);
+         nfts[_index].powerValue++;
+         playerpowervalue[msg.sender]++;    
+     }
+```
+
+The `UpgradeNFT()` function allows a player to upgrade the power value of their NFT. This function takes an `index` as a parameter which is used to specify the NFT to be upgraded. The function requires that the sender of the transaction is the owner of the NFT. If this requirement is met, the function transfers 0.5 Celo from the owner of the nft to the contract owner and increases the power value of the NFT and the player. 
+
+```solidity
+ function canSwallowNFT(address _address, uint _index) public view returns(bool){
+        if(playerpowervalue[_address] > playerpowervalue[nfts[_index].owner]){
+            return true;
+        }else{
+            return false;
+        }
+    }
+```
+
+The `CanSwallowNFT()` function returns true if the power value of the attacker is greater than the power value of the owner of the NFT. The function takes an `address` and an `index` as parameters. The address is used to specify the attacker and the index is used to specify the NFT to be swallowed. The function returns true if the power value of the attacker is greater than the power value of the owner of the nft. 
```

Comment:
The `CanSwallowNFT()` function returns true if the power value of the attacker is greater than the power value of the owner of the NFT. The function takes an `address` and an `index` as parameters. The address is used to specify the attacker, and the index determines the NFT to be swallowed. The function returns true if the power value of the attacker is greater than the power value of the owner of the nft.